File: mobile/modules/ikkyee-local-storage/ios/IkkyeeLocalStorageModule.swift
import ExpoModulesCore
import Foundation

public final class IkkyeeLocalStorageModule: Module {
  public func definition() -> ModuleDefinition {
    Name("IkkyeeLocalStorage")

    AsyncFunction("getDatabaseDirectoryObservation") { () -> [String: String] in
      let fileManager = FileManager.default
      guard let trustedRoot = fileManager.urls(
        for: .applicationSupportDirectory,
        in: .userDomainMask
      ).first else {
        throw IkkyeeLocalStorageError("Application Support directory is unavailable")
      }
      guard let bundleId = Bundle.main.bundleIdentifier, bundleId == expectedBundleId else {
        throw IkkyeeLocalStorageError("Unexpected iOS bundle identifier")
      }

      try fileManager.createDirectory(
        at: trustedRoot,
        withIntermediateDirectories: true
      )
      var databaseDirectory = trustedRoot.appendingPathComponent("ikkyee-local", isDirectory: true)
      try fileManager.createDirectory(
        at: databaseDirectory,
        withIntermediateDirectories: true
      )

      var resourceValues = URLResourceValues()
      resourceValues.isExcludedFromBackup = true
      try databaseDirectory.setResourceValues(resourceValues)
      let observedValues = try databaseDirectory.resourceValues(
        forKeys: [.isExcludedFromBackupKey]
      )
      guard observedValues.isExcludedFromBackup == true else {
        throw IkkyeeLocalStorageError("iOS did not verify backup exclusion")
      }

      return [
        "platform": "ios",
        "expectedBundleId": expectedBundleId,
        "adapterBundleId": bundleId,
        "trustedRootUri": canonicalDirectoryUri(trustedRoot),
        "databaseDirectoryUri": canonicalDirectoryUri(databaseDirectory),
        "trustedRootKind": "application-support",
        "verification": "native-adapter-observed",
        "nativeBackupExclusion": "verified"
      ]
    }
  }
}

private let expectedBundleId = "com.ikkyee.mobile"

private func canonicalDirectoryUri(_ url: URL) -> String {
  url.standardizedFileURL.absoluteString.trimmingCharacters(
    in: CharacterSet(charactersIn: "/")
  )
}

private struct IkkyeeLocalStorageError: Error, CustomStringConvertible {
  let description: String

  init(_ description: String) {
    self.description = description
  }
}

File: mobile/modules/ikkyee-local-storage/android/src/main/java/expo/modules/ikkyeelocalstorage/IkkyeeLocalStorageModule.kt
package expo.modules.ikkyeelocalstorage

import android.net.Uri
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

class IkkyeeLocalStorageModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("IkkyeeLocalStorage")

    AsyncFunction("getDatabaseDirectoryObservation") {
      val context = requireNotNull(appContext.reactContext) {
        "Ikkyee local storage requires an active application context"
      }
      val applicationId = context.packageName
      require(applicationId == EXPECTED_APPLICATION_ID) {
        "Unexpected Android application id"
      }

      val trustedRoot = context.noBackupFilesDir.canonicalFile
      val databaseDirectory = File(trustedRoot, "ikkyee-local").canonicalFile
      require(databaseDirectory.parentFile == trustedRoot) {
        "Database directory escaped the Android no-backup root"
      }
      require(databaseDirectory.isDirectory || databaseDirectory.mkdirs()) {
        "Unable to create the local photo database directory"
      }

      mapOf(
        "platform" to "android",
        "expectedApplicationId" to EXPECTED_APPLICATION_ID,
        "adapterApplicationId" to applicationId,
        "trustedRootUri" to Uri.fromFile(trustedRoot).toString(),
        "databaseDirectoryUri" to Uri.fromFile(databaseDirectory).toString(),
        "trustedRootKind" to "no-backup-files",
        "verification" to "native-adapter-observed"
      )
    }
  }

  private companion object {
    const val EXPECTED_APPLICATION_ID = "com.ikkyee.mobile"
  }
}
